File: app/src/main/java/com/willchou/dapenti/model/DaPenTiCategory.java
package com.willchou.dapenti.model;

import android.util.Pair;

import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class DaPenTiCategory {
    static private final String TAG = "DaPenTiCategory";

    private String categoryName;
    private URL categoryUrl;

    public List<DaPenTiPage> pages = new ArrayList<>();

    public interface onCategoryPrepared {
        void onCategoryPrepared(int index);
    }

    public onCategoryPrepared categoryPrepared = null;

    DaPenTiCategory(Pair<String, URL> p) {
        categoryName = p.first;
        categoryUrl = p.second;
    }

    public boolean initiated() { return !pages.isEmpty(); }
    public String getCategoryName() { return categoryName; }

    public void setPages(List<Pair<String, URL>> pair, boolean fromDatabase) {
        pages.clear();
        Database database = Database.getDatabase();
        for (Pair<String, URL> p : pair) {
            pages.add(new DaPenTiPage((p)));

            if (!fromDatabase && database != null)
                database.addPage(categoryName, p.first, p.second.toString());
        }

        if (categoryPrepared != null)
            categoryPrepared.onCategoryPrepared(pages.size() - 1);
    }

    public void preparePages(boolean fromWeb) {
        List<Pair<String, URL>> subItemPair = new ArrayList<>();

        boolean fromDatabase = true;
        Database database = Database.getDatabase();
        if (!fromWeb && database != null)
            database.getPages(categoryName, subItemPair);

        if (subItemPair.isEmpty()) {
            //String ss = "div > ul > li > a";
            String ss = "li>a[href^='more.asp?name='],span>a[href^='more.asp?name=']";
            subItemPair = DaPenTi.getElementsWithQuery(categoryUrl, ss);

            fromDatabase = false;
        }

        setPages(subItemPair, fromDatabase);
    }
}

File: app/src/main/java/com/willchou/dapenti/model/DaPenTiCategory.kt
package com.willchou.dapenti.model

import android.util.Log
import android.util.Pair
import java.net.URL
import java.util.*

class DaPenTiCategory internal constructor(p: Pair<String, URL>) {
    companion object {
        private const val TAG = "DaPenTiCategory"
    }

    val categoryName: String = p.first
    private val categoryUrl: URL = p.second

    var pages: MutableList<DaPenTiPage> = ArrayList()

    fun initiated(): Boolean {
        return !pages.isEmpty()
    }

    private fun setPages(pageInfoList: List<Database.PageInfo>, fromDatabase: Boolean) {
        pages.clear()
        val database = Database.database

        for (pageInfo in pageInfoList) {
            var page = DaPenTi.daPenTi?.daPenTiPageMap?.get(pageInfo.pageTitle)
            if (page != null)
                pages.add(page)
            else {
                page = DaPenTiPage(pageInfo.pageTitle, pageInfo.pageUrl, pageInfo.isFavorite)
                DaPenTi.daPenTi?.daPenTiPageMap?.put(pageInfo.pageTitle, page)
                pages.add(page)
            }
        }

        if (!fromDatabase && database != null) {
            for (pageInfo in pageInfoList.reversed())
                database.addPage(categoryName, pageInfo.pageTitle, pageInfo.pageUrl.toString())
        }

        DaPenTi.notifyCategoryChanged(categoryName)
    }

    fun preparePages(fromWeb: Boolean) {
        val pageInfoList: MutableList<Database.PageInfo> = ArrayList()

        var fromDatabase = true
        val database = Database.database
        if (!fromWeb && database != null)
            database.getPages(categoryName, pageInfoList)

        if (pageInfoList.isEmpty()) {
            //String ss = "div > ul > li > a";
            val ss = "li>a[href^='more.asp?name='],span>a[href^='more.asp?name=']"
            val subItemPair = DaPenTi.getElementsWithQuery(categoryUrl, ss)

            if (subItemPair.isEmpty()) {
                Log.d(TAG, "Unable to fetch from web: $categoryUrl")
                DaPenTi.notifyCategoryChanged(categoryName)
                return
            }

            for (p in subItemPair) {
                val favorite = Database.database?.getPageFavorite(p.first)
                val pi = Database.PageInfo(p.first, p.second,favorite != null && favorite)
                pageInfoList.add(pi)
            }

            fromDatabase = false
        }

        setPages(pageInfoList, fromDatabase)
    }
}
